libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -199,6 +203,35 @@ size_t WaveClip::GetAppendBufferLen() const
    return mSequences[0]->GetAppendBufferLen();
 }
 
+void WaveClip::OnProjectTempoChange(
+   const std::optional<double>& oldTempo, double newTempo)
+{
+   if (!mRawAudioTempo.has_value())
+      // When we have tempo detection ready (either by header-file
+      // read-up or signal analysis) we can use something smarter than that. In
+      // the meantime, use the tempo of the project when the clip is created as
+      // source tempo.
+      mRawAudioTempo = oldTempo;
+
+   if (oldTempo.has_value())
+   {
+      const auto ratioChange = *mProjectTempo / newTempo;
+      mSequenceOffset *= ratioChange;
+      mTrimLeft *= ratioChange;
+      mTrimRight *= ratioChange;
+   }
+   mProjectTempo = newTempo;
```

Comment:
Reviewing the fifth commit.  If I read the function above in isolation, I see there might be a crash if when `oldTempo.has_value()` but `mProjectRatio` does not.

I had to reread other code outside of WaveClip to convince myself the crash won't happen (`WaveTrack::InsertClip`).

I would prefer you just crash-proof the function by changing the numerator  to
`mProjectTempo.value_or(*oldTempo)`